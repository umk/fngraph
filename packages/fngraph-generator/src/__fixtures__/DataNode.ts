import DataNode from '../DataNode'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
