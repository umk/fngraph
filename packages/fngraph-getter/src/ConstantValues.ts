import { DataRecord } from '@fngraph/data'
import {
  CONTEXT_GROUP_INVERSION,
  Falsy,
  GeneratorValue,
  Getter,
  GetterFactory,
} from '@fngraph/generator'

import { createIteration } from './IterationFunction'

type ConstantValues<V extends DataRecord> = Array<V | Falsy> | V | Falsy

export function createConstant<V extends DataRecord>(
  source: ConstantValues<V>,
): GetterFactory<never, V> {
  return function (_incoming, outgoing): Getter {
    const iterationF = createIteration(outgoing)
    return async function* (
      _properties,
      records,
      context,
      getRecord,
      invert,
    ): AsyncGenerator<GeneratorValue> {
      const iteration = iterationF(context, invert)
      for await (const current of records) {
        context.advanceToValueGroup(current)
        const base = getRecord(current.contexts)
        if (base) {
          const valuesArr = Array.isArray(source) ? source : [source]
          if (!invert) {
            yield* iteration(current, base, valuesArr.filter(Boolean) as Array<V>)
          } else {
            if (!valuesArr.some(Boolean)) {
              yield* iteration(current, base, CONTEXT_GROUP_INVERSION as V)
            }
          }
        }
      }
    }
  }
}

export default ConstantValues
