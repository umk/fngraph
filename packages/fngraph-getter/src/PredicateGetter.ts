import { DataRecord } from '@fngraph/data'
import { GetterFactory } from '@fngraph/generator'

import { createOneToOne } from './OneToOneGetter'

type PredicateGetter<P extends DataRecord> = (record: P) => Promise<unknown>

export function createPredicate<P extends DataRecord>(
  source: PredicateGetter<P>,
): GetterFactory<P, Record<string, never>> {
  return createOneToOne(async (record) => ((await source(record as P)) ? {} : undefined))
}

export default PredicateGetter
