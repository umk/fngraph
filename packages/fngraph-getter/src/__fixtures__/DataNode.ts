import { DataNode } from '@fngraph/generator'

export function createDataNode(overrides?: Partial<DataNode>): DataNode {
  return {
    incoming: [],
    outgoing: [],
    getter: jest.fn(),
    priority: 0,
    invert: false,
    getProperties: () => [],
    ...overrides,
  }
}
