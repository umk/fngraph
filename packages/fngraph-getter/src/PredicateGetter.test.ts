import { DataRecord, DeclarationID } from '@fngraph/data'
import { contextAsRecord, createDataRecordGenerator, recordAsContext } from '@fngraph/generator'

import { createConstant } from './ConstantValues'
import { createPredicate } from './PredicateGetter'
import { createDataNode } from './__fixtures__'

describe('createPredicate', () => {
  const declarationA = 'declaration A' as DeclarationID
  const declarationB = 'declaration B' as DeclarationID
  it('iterates through getter records', async () => {
    const values = new Array(25).fill(0).map((_, i) => ({ [declarationA]: i }))
    const nodeA = createDataNode({
      outgoing: [declarationA],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const nodeB = createDataNode({
      incoming: [declarationA],
      getter: createPredicate((record) => {
        const value = record[declarationA] as number
        return Promise.resolve(value % 2 === 0)
      })(contextAsRecord, recordAsContext),
    })
    const generator = createDataRecordGenerator([nodeA, nodeB], [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
  })
  it('permutates records with another node', async () => {
    /**
     * Must result in the following evaluation sequence:
     *
     *   (B) <- (C) <- (A)
     *
     *        - or -
     *
     *   (B) <- (A) <- (C)
     */
    const values = new Array(25).fill(0).map((_, i) => ({ [declarationA]: i }))
    const nodeA = createDataNode({
      outgoing: [declarationA],
      getter: createConstant(values)(contextAsRecord, recordAsContext),
    })
    const getter = jest.fn((record) => {
      const value = record[declarationA] as number
      return Promise.resolve(value % 3 === 0)
    })
    const nodeB = createDataNode({
      incoming: [declarationA],
      getter: createPredicate(getter)(contextAsRecord, recordAsContext),
    })
    const nodeC = createDataNode({
      outgoing: [declarationB],
      getter: createConstant([{ [declarationB]: 10 }, { [declarationB]: 20 }])(
        contextAsRecord,
        recordAsContext,
      ),
    })
    const generator = createDataRecordGenerator([nodeA, nodeB, nodeC], [])
    const records: Array<DataRecord> = []
    for await (const record of generator()) records.push(record)
    expect(records).toMatchSnapshot()
    expect(getter).toHaveBeenCalledTimes(25)
  })
})
