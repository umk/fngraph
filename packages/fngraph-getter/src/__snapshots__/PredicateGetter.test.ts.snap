// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPredicate iterates through getter records 1`] = `
[
  {
    "declaration A": 0,
  },
  {
    "declaration A": 2,
  },
  {
    "declaration A": 4,
  },
  {
    "declaration A": 6,
  },
  {
    "declaration A": 8,
  },
  {
    "declaration A": 10,
  },
  {
    "declaration A": 12,
  },
  {
    "declaration A": 14,
  },
  {
    "declaration A": 16,
  },
  {
    "declaration A": 18,
  },
  {
    "declaration A": 20,
  },
  {
    "declaration A": 22,
  },
  {
    "declaration A": 24,
  },
]
`;

exports[`createPredicate permutates records with another node 1`] = `
[
  {
    "declaration A": 0,
    "declaration B": 10,
  },
  {
    "declaration A": 3,
    "declaration B": 10,
  },
  {
    "declaration A": 6,
    "declaration B": 10,
  },
  {
    "declaration A": 9,
    "declaration B": 10,
  },
  {
    "declaration A": 12,
    "declaration B": 10,
  },
  {
    "declaration A": 15,
    "declaration B": 10,
  },
  {
    "declaration A": 18,
    "declaration B": 10,
  },
  {
    "declaration A": 21,
    "declaration B": 10,
  },
  {
    "declaration A": 24,
    "declaration B": 10,
  },
  {
    "declaration A": 0,
    "declaration B": 20,
  },
  {
    "declaration A": 3,
    "declaration B": 20,
  },
  {
    "declaration A": 6,
    "declaration B": 20,
  },
  {
    "declaration A": 9,
    "declaration B": 20,
  },
  {
    "declaration A": 12,
    "declaration B": 20,
  },
  {
    "declaration A": 15,
    "declaration B": 20,
  },
  {
    "declaration A": 18,
    "declaration B": 20,
  },
  {
    "declaration A": 21,
    "declaration B": 20,
  },
  {
    "declaration A": 24,
    "declaration B": 20,
  },
]
`;
